From 8f4c189d7009bf0ec8485d56c687b8552fa6225d Mon Sep 17 00:00:00 2001
From: Ning Sun <ning.sun@intel.com>
Date: Mon, 18 May 2015 11:44:31 -0700
Subject: This is a fix for Gentoo Hardened

which uses the GRSecurity and PaX patch sets on top of the mainstream
linux kernel.  Disables PCID just before disabling paging in the
shutdown process will prevent the machine to reboot in the case of tboot
was not exited properly when with PCID enabled.

Signed-off-by Jason Zaman <jason@perfinion.com>
Signed-off-by Ning Sun <ning.sun@intel.com>
---
 Makefile                  |    2 +-
 tboot/common/shutdown.S   |    6 ++++++
 tboot/include/processor.h |    1 +
 3 files changed, 8 insertions(+), 1 deletions(-)

diff --git a/Makefile b/Makefile
index 3e118e0..efb17e3 100644
--- a/Makefile
+++ b/Makefile
@@ -122,7 +122,7 @@ help :
 	@echo '  world            - clean everything'
 	@echo ''
 	@echo 'Cleaning targets:'
-	@echo '  clean            - clean sboot and tools'
+	@echo '  clean            - clean tboot and tools'
 	@echo '  distclean        - clean and local downloaded files'
 	@echo ''
 	@echo '  uninstall        - attempt to remove installed tools'
diff --git a/tboot/common/shutdown.S b/tboot/common/shutdown.S
index af97513..fbcaf26 100644
--- a/tboot/common/shutdown.S
+++ b/tboot/common/shutdown.S
@@ -34,6 +34,12 @@
  */
 
 compat_mode_entry:
+        /* Disable PCID */
+        movl %cr4, %eax
+        andl $~CR4_PCIDE, %eax
+        movl %eax, %cr4
+
+
 	/* Disable paging and therefore leave 64 bit mode. */
         movl %cr0, %eax
         andl $~CR0_PG, %eax
diff --git a/tboot/include/processor.h b/tboot/include/processor.h
index 015993f..4e24ec3 100644
--- a/tboot/include/processor.h
+++ b/tboot/include/processor.h
@@ -92,6 +92,7 @@
 #define CR4_XMM 0x00000400 /* enable SIMD/MMX2 to use except 16 */
 #define CR4_VMXE 0x00002000/* enable VMX */
 #define CR4_SMXE 0x00004000/* enable SMX */
+#define CR4_PCIDE 0x00020000/* enable PCID */
 
 #ifndef __ASSEMBLY__
 
-- 
1.7.1

